### ISSUE TITLE
Create doc on how to support ESM and CommonJS when publishing modules

### ISSUE THREAD

**Issue by mhdawson** [nodejs/package-maintenance#542]
From discussion at the last collaborator summit. 

This would help as the suggestion was that there were some easy ways to do this.

**Comment by lemanschik** [comment]
@ljharb for example that you should prefer only namedExports from esm and transpile to CJS there are a lot of clashes when it comes to typescript because it ads a extra resolve matrix that is not compat

a great example are .d.ts files which use export = Module all that can get avoided if we do only namedExports by default end avoid assigning anything else then string propertys to a Object on module.exports

so never do module.exports = function. that also saves a lot of tooling quircs i can write whole books about this cross issues that come up as soon as you bundle more then 1 single code base. 

at present for example i am struggling on puppeteer and devtools-protocol also vscode makes a lot of trouble in general all the big code bases that have typescript in it are relativ blocked from ESM migration.

**Comment by ljharb** [comment]
@lemanschik i strongly disagree with that; a default export is what a module *is*; named exports are what module *has*, and most modules should *be* something. That’s a stylistic choice and not something node should be weighing in on. I, and the Airbnb style guide ftr, say that you should almost *always* do module.exports = function.

**Comment by lemanschik** [comment]
@ljharb they say so because of compat i am working since ages with ESM only and can report back that no one did produce correct ESM Builds that load in a correct ESM Environment. (Other then NodeJS)

**Comment by ljharb** [comment]
I have a number of projects that work perfectly compatibly with a build process, so I’m not sure about those reports.

**Comment by lemanschik** [comment]
@ljharb your again correct with build tools which correct the imports everything works at the end but the Nativ code that should work untranspiled and should even get patched at runtime does not work. 

thats why i got rollup and typescript contributor only because of such issues.

**Comment by lemanschik** [comment]
Maybe it helps to show a true isomorphic cjs ESM module that uses a global 

Load Isomorphic CJS that uses only dynamic import and assigns to global as ESM and CJS share the same global this works. 

```js
globalThis.module = { exports: {}}
import('').then(_m=>(v => {
 const v = globalThis.module.exports
 delete globalThis.module.exports
 return v
});
```

**Comment by lemanschik** [comment]
This pattern is also nice when you work with CJS ESM NodeJS


my_esm.js
```js
const moduleName = { fnName() {} }
export { moduleName }
```

my_cjs_wrapper
```js

const moduleName = {
   async fnName() {
    return (await(import('./my_esm.js')).fnName()
  }
}
```

**Comment by lemanschik** [comment]
Maybe i should also open some eyes for how big this problem is:

- [ ] https://github.com/microsoft/TypeScript/issues?q=is%3Aissue+is%3Aopen+moduleResolution

this is in general a 1:1 Representation of the real nightmare that is under the hood of bundlers which abstract this reality. 

to be fair most of that is caused by NPM CJS TYPESCRIPT. Pure ESM always works and thats why it is so importent to go ESM only in the source code and then generate obscured builds but not code boscured and generate 

even more obscured code.

Companys like Blomberg did invent tooling stacks to unify depdendencys and maintaing them is hell

I did by the way audit over 8k NPM Modules by Hand in the last years and that are only the once that where so interristing that i did even forked them to play around.

**Comment by lemanschik** [comment]
## Complex educational problems
loop, async, globalThis aka Context, Scopes, incremental js upgrading via "use strict" directive

all this maybe sounds short and clear but all that is a complex matrix and it has effects on the behavior of objects and everything is a object in ECMAScript/JS

**Comment by lemanschik** [comment]
## This should get a list of problems maybe 

- [ ] https://github.com/microsoft/playwright-java <= ships whole NodeJS to get Playwright running which has less then 5mb raw code in ESM none minified all comments included. so shipping graal-node (A NodeJS Fork to run nodejs ECMAscript on GraalVM)

**Comment by lemanschik** [comment]
## Update
- [ ] https://github.com/nodejs/modules/issues/351 when solved so a plugin system is agreed on which could be soon as import require flags got added for preloaders before frozen intrinsics that enables using the same Loader for CJS ESM. Workers, And even Browser Contexts. This will allow us to author a guide like

## Package Gold Standard

- Author your code
- Write load(resolvedId), resolveId(importSpecifier, importee) hooks 
- Write your code 
- WiP ... aggree on something to define what to us and how to use the loader for the written code.

**Comment by lemanschik** [comment]
@ljharb i am happy that you ask for example it makes total new forms of builds more enjoyable and compose able at present we have overall good results with opening component based ipc specifiers inside chromium and v8 it self. this leads to stuff like running the monaco editor as chrome extension directly then using workspaces and overrides and then directly use unmodified directly loaded modules out of the npm ecosystem in a way that i can patch and replace them or use single functions of them to compose something total diffrent

It is simply expressiv when i can write

```js
const reused = import('anything').then(any=> Object.assign(any, { addOrpatch }))
```

**Comment by ljharb** [comment]
@lemanschik @frank-dspeed i'm not sure that answers the question. CJS is equally composable; enjoyable is subjective. I agree that a build process remains required, but I haven't seen an advisable "no build process" setup anywhere yet, so i'm not sure why that's important.

**Comment by lemanschik** [comment]
@ljharb you will see it in action at present we use that all over simply via userland loaders but we would benefit when npm users would create packages in a way that we do not need to handle nodejs logic so we can depend on a nodejs indidpendent logic for the resolve. that allows us to do content based hashing and cascasing builds 

so we can hot replace modules in large builds see it as fast distributed hot module reload helper in general we at present worked a lot on tooling to create indipendent specifier resolvers all over the ecosystem.

When NodeJS and the NPM System would agree on such thing we simply can save a lot of build time energie and frustrations for millions of developers. 

For example if some one patches a single module in a stack of millions where some do depend of it that helps a lot.

**Comment by lemanschik** [comment]
maybe the best description what all that does solve is it reduces the need to implement own resolver implementations inside every package manager and stuff like typescript and vscode electron can share the resolve logic cross context.

and also our chromium ipc can share the resolve logic via remote pipe stdio imagin tesing your builds directly in the browser no need for input output directly fix bugs and hit run again in the same view.

**Comment by ljharb** [comment]
Hot reloading all modules generically isn't a thing that's possible without engine cooperation, due to the nature of JavaScript itself, and node would require support from v8 to do it. Either way, this issue's document should be written to fit actual reality, not aspirational reality.

**Comment by lemanschik** [comment]
@ljharb ok then your correct lets start with a doc that points out your observations always bundle to CJS and in some months we will revisit that so no one gets hurt by writing that up for no need.

**Comment by lemanschik** [comment]
lets add the recommendation to create a ESM wrapper

wrapper.js
```js
module.exports = import('./esm.mjs')
```

**Comment by ljharb** [comment]
That would be a harmful recommendation, because then you'd be exporting a Promise for a Module object. ESM wrappers don't add value.

**Comment by lemanschik** [comment]
hmm ok i guess as you sayed transpil everything there will be no edgecase where something needs to stay esm?

like scripts that use the import.meta.url on dynamic load and parse url parms to return a conditional module?

**Comment by ljharb** [comment]
Correct, that's my understanding. The only thing is top-level await, and i've seen no reasonable use cases for that besides an application entrypoint.

**Comment by lemanschik** [comment]
@ljharb i invested into some research on HTTP Archive it is a little collection of the whole web which is queryable for code. we use it often to evaulate deprecations web wide and collect feature usage. 

Your correct that you not saw much and will not see much at present. As it is blocked all over by other factors.

At scale ESM Makes sense but i agree that out of normal people view and normal small crm project sizes all this does not matter. While companys lke Facebook and Google and blomberg would save N*N*N*N Money overall. 

the most importent details are small but they add up and add up. About the v8 support that is there unreferenced modules get thrown out of the code cache. For example we have a small project a PoC Demo it shows what we can do with raw v8 which we use to demo why we do components at all.

i present to you: https://github.com/just-js/just a raw v8 minimal runtime that builds it self and ranks 1 on reputable benchmarks like this. https://www.techempower.com/benchmarks/#section=data-r21&test=composite 

with that supplyed benchmark suite: https://github.com/just-js/techempower

and there for example we have also no require and not a need for anything like it. We would at scale simply save tons of CO2. But i guess for 99% of people your right and it makes no real effect for them until adoption reaches 60%+

**Comment by vostrik** [comment]
Hi there 👋 
Your discussion is exciting and informative.
I see here 2 points of view.

Can we add drafts to this issue from both opinions in format like this:

1. Original demo JavaScript code
2. Export notation for CJS and ESM
3. Settings samples for popular bundlers (webpack, rollup, parcel, maybe more)

I. e. add `examples` folder to our repository and 2 samples which we can test in GitHub Codespaces and discuss about it. 

It can be more evident for us and for the community. 
What do you think?

**Comment by lemanschik** [comment]
@vostrik the thing is i am working on using no codespaces any more in fact i do not use them they are waist i try to compress the knowleg

- [ ] https://github.com/lemanschik/simplicity-made-easy

once the knowleg transport is there every one understands that he can work directly in the browser with the ESM modules also there will be a example of using git commits directly as modules as git commits are text base modules that are compose able out of references. Even cherry picking 

No infrastructure is needed to build and run code no servers are needed every one can be his own domain. but it is hard to educate the people there is a big gap so i need to write whole books of examples as also translate existing docs of core ecosystem packages to show the people that they can simply throw the existing kubernetes infrastructure away.

and directly execute the code without roundtrips which leads to faster iteration and debugging

**Comment by lemanschik** [comment]
the main problem here is static vs dynamic CJS returns DYNAMIC results and i want static to do incremental improvements.

**Comment by lemanschik** [comment]
the above bespoken turns your builds into *nix like builds like the nix package manager. you can google it. Thats the same concept in ecmascript the nix store is your ESM cache.

**Comment by lemanschik** [comment]
so mainly the above little module examples implement near everything that git offers at runtime branch fork code deduplication.. content deduplication. and so on and so on. stable versioning.

**Comment by lemanschik** [comment]
Long story short not going ESM only Blocks Web 4.0 and whole internet evolution as also every developer that depends on this CJS stuff. They are all depending on legacy tech as sayed all this got planned since ages i work on that since 30+ years.

**Comment by lemanschik** [comment]
the biggest part of the ecosystem or lets say one of the biggest

https://github.com/sindresorhus/meta/discussions/15

he maintains a lot of packages and we can be thankful that he did the esm move and so forces rethinking he is a legend!

**Comment by vostrik** [comment]
@lemanschik thank you so much for this detailed response.
I've just read all of your replies including tiny book and small discussion in TS repo by links above.
I appreciate it.

But I also have 2 theses:
1. I really think that it would be more effective to discuss about new module approaches in [Modules Team](https://github.com/nodejs/modules).
2. Main topic for this discussion is `Create doc on how to support ESM and CommonJS when publishing modules`. That's why we should focus on current ecosystem state and document it. Of course with simplicity criteria as u described.

Let me repeat my suggestion: 

> Can we add drafts to this issue from both opinions in format like this:
> 1. Original demo JavaScript code
> 2. Export notation for CJS and ESM
> 3. Settings samples for popular bundlers (webpack, rollup, parcel, maybe more)
> I. e. add examples folder to our repository and 2 samples which we can test in GitHub Codespaces and discuss about it.

@ljharb, @mhdawson tell me please, what do you think about my propose and about maybe(!) a little bit off-topic vibe from previous messages.

P. S. @lemanschik I really appreciate your thoughts and even ready to participate in PoC of your idea if needed.

**Comment by lemanschik** [comment]
@vostrik you do not got the full picture https://stackblitz.com/edit/node-ljbwfd?file=index.js

companys like stackblitz for example did already complet wasm builds of NODEJS in the browser.

Nodejs community is to slow now committee needed anymore we are all already free i will Update vscode code-oss and other editors to produce the correct result out of the box

this also replaces packages as a whole! it deprecates kubernetes and docker.. Even the people of stackblitz are far away from the correct solution but more near then nodejs it self

**Comment by lemanschik** [comment]
The most big problem is there is no real point to start as all is done since 2018 but now the ecosystem goes wild but i do not need to care so much about that as cloudflare and the other big providers also do not care for the fact that the world is now a open shell i will simply do what all do after i did full public disclosure which i realy did i mean

you can buy on steam a game with the name "turing complete" it cost 20 dollar and shows you how to build a PC then use ECMAScript to follow what you learned and your done.  you took over the internet.

ok that was the short story but there are no Modular Web Application firewalls and package management systems for the most people as they do not understand how to use the fundamentals. 

All protection inherent by the usage of isolated files and network as also inside that a package manager that checks file integrity all that is not there on the net

so if you skip the local stack you got exactly a Open Shell like it was 30 years before when you found a open ftp that supports command execution.

**Comment by ljharb** [comment]
@lemanschik can you please stop posting dozens of essays? The thread is already impossibly long to read.

**Comment by lemanschik** [comment]
@ljharb your correct sorry but i will clean that up i swear!

**Comment by ljharb** [comment]
That’s not likely to happen, and either way it’s entirely off topic for this thread.

This is about documenting what *is*, not trying to imagine a different reality.

**Comment by lemanschik** [comment]
@ljharb that is the current case but no one uses it please bring up blockers.

**Comment by lemanschik** [comment]
but your correct i will delete my interventions here it was to noisy maybe deprecating nodejs is the right thing to do and implement a browser nodejs api shim that works even with headless mode. thanks for inspiration

**Comment by mhdawson** [comment]
@ljharb doc where you can fill in the approach you discussed - https://docs.google.com/document/d/1NIYPxF4OsVBX3ES9Gx48pomZlWtQNG1NIiY9aRSpp8k/edit

**Comment by lholmquist** [comment]
Found this on the node.js docs:  https://nodejs.org/api/packages.html#dual-commonjses-module-packages

Does this cover what we wanted?

**Comment by frank-dspeed** [comment]
@lholmquist it was not what we search for but it is today maybe the thing we search for let me explain:

Out of Historical Reason there where 1x Export per Package JSON then later with es2015 the package json got the import / export field which can store multiple exports. 

This exports have also support for nested conditional exports thats the part that your posting. 

So the Field is only supported by current tooling every tooling today supports them as far as i know. If not support can be added in less then a day.

the --condition flag lets you choose one set of imports/exports.

Also i can announce the landing of require esm inside NodeJS 22+ 

````js
require('es-module.js')
````

works under a flag if the ESM does not contain top level await at last.

**Comment by lemanschik** [comment]
Here's the corrected version of your text:

---

I would propose the following path forward now:

- Create ESM-only packages and only use the module field. Call it a day.
- Use the `-r esm` flag to require the ESM NPM module to activate user-land support for `require(esm-module.js)` with additional support for top-level await as a migration path.
- From Node.js 22+ onward, migrate to `--experimental-require-module` and `-r esm` until the other flag works.
- From Node.js 22.x+ when the experimental flag gets an opt-out flag, call it a day. We will have resolved the ambiguity. Top-level import can import ESM and CJS code, and `require` can do the same. While everything should get transpiled to ESM for analyzability and tooling.

---

**Comment by bjohansebas** [comment]
fixed by https://github.com/nodejs/package-examples and https://nodejs.org/en/learn/modules/publishing-a-package
